node/src/components/consensus/protocols/highway/round_success_meter.rs

```diff
@@ -0,0 +1,146 @@
+    }
+
+    /// Registers a proposal within this round - if it's finalized within the round, the round will
+    /// be successful.
+    pub fn new_proposal(&mut self, proposal_h: C::Hash, timestamp: Timestamp) {
+        // only add proposals from within the current round
+        if timestamp.millis() >= self.current_round_id
+            && timestamp.millis() - self.current_round_id < self.current_round_length
+        {
+            self.proposals.push(proposal_h);
+        }
+    }
+
+    /// If the current timestamp indicates that the round has ended, checks the known proposals for
+    /// a level-1 summit.
+    /// If there is a summit, the round is considered successful. Otherwise, it is considered
+    /// failed.
+    /// Next, a number of last rounds are being checked for success and if not enough of them are
+    /// successful, we return a higher round exponent for the future.
+    /// If the exponent shouldn't grow, and the round ID is divisible by a certain number, a lower
+    /// round exponent is returned.
+    pub fn calculate_new_exponent(&mut self, state: &State<C>) -> u8 {
+        let now = Timestamp::now();
+        // if the round hasn't finished, just return whatever we have now
+        if now.millis().saturating_sub(self.current_round_id) < self.current_round_length {
+            return self.new_exponent();
+        }
+
+        let current_round_index = self.current_round_id / self.current_round_length;
+        let new_round_index = now.millis() / self.current_round_length;
+
+        if mem::replace(&mut self.proposals, Vec::new())
+            .into_iter()
+            .any(|proposal| self.check_proposals_success(state, &proposal))
+        {
+            self.rounds.push_front(true);
+        } else {
+            self.rounds.push_front(false);
+        }
+
+        // if we're just switching rounds and more than a single round has passed, all the
+        // rounds since the last registered round have failed
+        for _ in 0..(new_round_index - current_round_index - 1) {
+            self.rounds.push_front(false);
+        }
+
+        self.current_round_id = new_round_index * self.current_round_length;
+
+        self.clean_old_rounds();
+
+        let new_exp = self.new_exponent();
+
+        if new_exp != self.current_round_exp {
+            self.change_exponent(new_exp, now);
+        }
+
+        new_exp
+    }
+
+    fn clean_old_rounds(&mut self) {
+        while self.rounds.len() > NUM_ROUNDS_TO_CONSIDER {
+            self.rounds.pop_back();
+        }
+    }
+
+    fn count_failures(&self) -> usize {
+        self.rounds.iter().filter(|&success| !success).count()
+    }
+
+    fn new_exponent(&self) -> u8 {
+        let current_round_index = self.current_round_id / self.current_round_length;
+        if self.count_failures() > MAX_FAILED_ROUNDS {
+            self.current_round_exp + 1
+        } else if current_round_index % ACCELERATION_PARAMETER == 0
```

Comment:
So acceleration and deceleration conditions are not symmetric?

Reply:
Yes. That's because we want the validators to speed up simultaneously if they are speeding up at all. Otherwise, a single validator speeding up will fail every other round, because others will be too slow to help it finalize blocks faster.

Reply:
Sure, but eventually others would join him. Is that a problem that he would fail to finalize every other round of his?

Reply:
> Sure, but eventually others would join him.

That's the problem - not necessarily. The validator in question would likely slow down again before others joined it (or even a bit after, but it would already have accumulated enough failures to decide to slow down). So then we would have validators accelerating and slowing down at random times.

Reply:
Ah, right. B/c `MAX_FAILURES=10` so validator that goes too fast will always slow down since it will have half of the rounds failed. Does it mean that we if we set `MAX_FAILURES > NUM_ROUNDS_TO_CONSIDER / 2` then we could allow them to speed up independently? Because on average only half of the rounds will fail to finalize.

Reply:
Theoretically yes... But this would also be undesirable, because what if all the validator are exactly 2 times faster than their limit? If they decelerated, they would be working exactly as fast as they can, but they wouldn't decelerate, because they would fail exactly half of the time, which with that threshold wouldn't be enough to make the decision.

Reply:
OK, you have convinced me :+1: For now. I still don't like how these values are chosen arbitrarily but we can tweak it in the future.